Attempt to support aggregate_double_metric field type in esql.

martijnvg opened the pull request:
This tries to add support for this field type by:
* Normalizing to the element type double.
* Adding a hint property to field attribute. The hint gets set by a few numeric aggregates.
* Using that hint in AggregateDoubleMetricFieldMapper#blockLoader(...) to pick the right subfield (min, max, sum or count).

This simplistic approach seems to work for min, max, sum and avg functions. Example: `FROM tsdb-1d | STATS avg(kubernetes.container.memory.usage.bytes)`

This is WIP with hacks and I didn't run tests yet, just playing around with a local ES instance with downsampled indices.
Opening this draft PR just to check whether the direction of this change makes sense.

Relates to #110649

martijnvg committed:
Attempt to support aggregate_double_metric field type in esql.

This tries to add support for this field type by:
* Normalizing to the element type double.
* Adding a hint property to field attribute. The hint gets set by a few numeric aggregates.
* Using that hint in AggregateDoubleMetricFieldMapper#blockLoader(...) to pick the right subfield (min, max, sum or count).

This simplistic approach seems to work for min, max, sum and avg functions.
Example: `FROM tsdb-1d | STATS avg(kubernetes.container.memory.usage.bytes)`

This is WIP and I didn't run tests yet, just playing around with a local ES instance with downsampled indices.

martijnvg commented on x-pack/plugin/mapper-aggregate-metric/src/main/java/org/elasticsearch/xpack/aggregatemetric/mapper/AggregateDoubleMetricFieldMapper.java:
```diff
@@ -510,6 +512,24 @@ public ValueFetcher valueFetcher(SearchExecutionContext context, String format)
+                case "min" -> delegateFieldType(Metric.min).blockLoader(blContext);
+                case "sum" -> delegateFieldType(Metric.sum).blockLoader(blContext);
+                case "count" -> {
+                    var fieldType = delegateFieldType(Metric.value_count);
```
This is one of the hacks. This makes sure to load double blocks, otherwise a class cast exception occurs in `SumDoubleAggregatorFunction.java:60`

dnhatn commented:
Thanks, Martijn. I think adding hints is a great way to limit the number of blocks we need to load instead of loading all blocks. However, I have some questions about the current approach:

1. Did you test this with `avg()`? I believe this approach may not return the correct value since it extracts only one block instead of two. For `avg`, we need two blocks: `sum` and `count`.

2. How does the approach handle queries using two different hints, for example:
   ```
   FROM tsdb-1d | STATS avg(kubernetes.container.memory.usage.bytes), max(kubernetes.container.memory.usage.bytes)
   ```
   
3. Should we support retrieving fields without aggregations? It's fine if we decide to postpone this:
   ```
   FROM tsdb-1d | kubernetes.container.memory.usage.bytes
   ```

martijnvg commented:
Thanks for taking a look Nhat!

> Did you test this with avg()? I believe this approach may not return the correct value since it extracts only one block instead of two. For avg, we need two blocks: sum and count.

I think it works. While I was working on this, I learned that `avg` function is a `SurrogateExpression` and that it translate `Div` function with a separate sum and count expression. The different hints for the `Sum` and `Count` trigger using different block loader in `AggregateDoubleMetricFieldMapper`.

> How does the approach handle queries using two different hints

That appears to be not working correctly. The result of the avg changes, when I switched between min and max functions. Looks like blocks / field attributes get reused, which shouldn't happen. I will look into this and see what can be done about this.

> Should we support retrieving fields without aggregations? It's fine if we decide to postpone this

Good question. That doesn't work with this change. Personally I think we don't to support this for now. I think the important part is that aggregate double metric field type works with the aggregate functions.

martijnvg committed:
fix yaml tests

martijnvg committed:
Merge remote-tracking branch 'es/main' into esql_aggregate_double_metric_field_type_1

martijnvg committed:
fix serialization versioning

martijnvg committed:
spotless

martijnvg committed:
Merge remote-tracking branch 'es/main' into esql_aggregate_double_metric_field_type_1

martijnvg committed:
fix compile error after merging in main branch

martijnvg committed:
update transport version

martijnvg committed:
update cluster features

martijnvg committed:
Merge remote-tracking branch 'es/main' into esql_aggregate_double_metric_field_type_1

martijnvg committed:
fixed compile error after merging main

martijnvg committed:
added esql.aggregate_metric_double node feature flag

martijnvg committed:
Merge remote-tracking branch 'es/main' into esql_aggregate_double_metric_field_type_1

martijnvg committed:
fix compile error after merging in main

martijnvg committed:
another iteration: move away from hint to make use the subfields of AggregateDoubleMetricFieldMapper.

AggregateDoubleMetricFieldMapper no longer hides its subfields. This allows esql to know about these fields, so that usage of the subfields is valid.

On the analyzer side, enrich field attributes with info that can be used by functions to pick right sub field.

martijnvg committed:
added todo

martijnvg committed:
adjust tests

martijnvg committed:
temp mute

martijnvg committed:
AggregateDoubleMetricFieldMapper no longer shares sub fields.
ESQL's Analyzer adds the AggregateDoubleMetricFieldMapper's sub fields manually.

martijnvg committed:
added required cluster feature to yml test

martijnvg committed:
added todo

martijnvg committed:
remove aggregate_double_metric field from 40_unsupported_types.yml,
because it is supported now.

martijnvg committed:
use random double for aggregate_metric_double literal

martijnvg committed:
move replacing of aggregate_metric_double field to sub field to `surrogate()` method and fix unit tests.

martijnvg committed:
Remove cluster feature requirement and instead use regex that allows aggregate_metric_double part to be optional in the regex

martijnvg committed:
Move aggregate double metric code out of FieldAttribute into seperate attribute classes.

martijnvg committed:
spotless

martijnvg committed:
simplify

martijnvg committed:
updates cluster feature

martijnvg committed:
update data type size.

dnhatn commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/aggregate/Avg.java:
```diff
@@ -73,6 +75,18 @@ public Expression surrogate() {
+            FieldAttribute sumField = aggregateDoubleMetricAttribute.getSumSubField();
+            sum = new Sum(s, sumField);
+            FieldAttribute countField = aggregateDoubleMetricAttribute.getValueCountSubField();
+            count = new Count(s, countField);
```
This should be `Sum(s, countField)` here?

dnhatn commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/aggregate/Count.java:
```diff
@@ -91,6 +94,9 @@ public DataType dataType() {
 
     @Override
     public AggregatorFunctionSupplier supplier(List<Integer> inputChannels) {
+        if (field() instanceof AggregateDoubleMetricSubAttribute) {
```
I think we can remove this?

martijnvg replied:
Good suggestion Nhat. I was able to remove it.

dnhatn commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/aggregate/Count.java:
```diff
@@ -125,8 +131,10 @@ public Expression surrogate() {
                 new Count(s, new Literal(s, StringUtils.WILDCARD, DataType.KEYWORD))
             );
+        } else if (field instanceof AggregateDoubleMetricAttribute aggregateDoubleMetricAttribute) {
+            return new Count(source(), aggregateDoubleMetricAttribute.getValueCountSubField());
```
I think this should be `Sum(source(), aggregateDoubleMetricAttribute.getValueCountSubField())` and we can remove the override `supplier` above?

dnhatn reviewed (commented):
This approach looks promising. However, we need to find a way to encapsulate the sub-fields and block loaders. Thanks @martijnvg

martijnvg committed:
Merge remote-tracking branch 'es/main' into esql_aggregate_double_metric_field_type_1

martijnvg committed:
Remove the change to `Count#supplier(...)` and instead in `Count#surrogate(...)` use `Sum` when encountering an `AggregateDoubleMetricAttribute` instance and use `Sum` in `Avg#surrogate(...)`.

martijnvg committed:
Merge remote-tracking branch 'es/main' into esql_aggregate_double_metric_field_type_1

martijnvg committed:
Merge remote-tracking branch 'es/main' into esql_aggregate_double_metric_field_type_1

martijnvg committed:
Replace AggregateDoubleMetricSubAttribute with normal FieldAttribute instances for the sub fields.

martijnvg committed:
Fold AggregateDoubleMetricAttribute into FieldAttribute.

martijnvg committed:
Merge remote-tracking branch 'es/main' into esql_aggregate_double_metric_field_type_1

martijnvg committed:
spotless
